File: minigit-3/py/commands/log.py
"""log command - Show commit history."""

import sys
from pathlib import Path
from datetime import datetime


def run(args: list[str]) -> int:
    """Show commit history."""
    from utils import find_repo_root
    from refs import resolve_ref, get_head_sha, list_branches
    from objects import read_commit

    repo_root = find_repo_root()
    if repo_root is None:
        print("error: not a minigit repository", file=sys.stderr)
        return 1

    # Parse arguments
    oneline = False
    show_all = False
    show_graph = False
    show_stat = False
    limit = None
    start_ref = None

    i = 0
    while i < len(args):
        arg = args[i]
        if arg == '--oneline':
            oneline = True
        elif arg == '--all':
            show_all = True
        elif arg == '--graph':
            show_graph = True
        elif arg == '--stat':
            show_stat = True
        elif arg == '-n' and i + 1 < len(args):
            try:
                limit = int(args[i + 1])
            except ValueError:
                pass
            i += 1
        elif not arg.startswith('-'):
            start_ref = arg
        i += 1

    # Get starting commits
    starting_shas = []

    if show_all:
        # Start from all branches
        for branch in list_branches(repo_root):
            sha = resolve_ref(repo_root, branch)
            if sha:
                starting_shas.append(sha)
    elif start_ref:
        sha = resolve_ref(repo_root, start_ref)
        if sha is None:
            print(f"error: unknown revision '{start_ref}'", file=sys.stderr)
            return 1
        starting_shas.append(sha)
    else:
        sha = get_head_sha(repo_root)
        if sha is None:
            print("error: no commits yet", file=sys.stderr)
            return 1
        starting_shas.append(sha)

    # Walk commit history
    visited = set()
    to_visit = list(starting_shas)
    commits = []

    while to_visit:
        sha = to_visit.pop(0)
        if sha in visited:
            continue
        visited.add(sha)

        try:
            commit = read_commit(repo_root, sha)
            commits.append((sha, commit))

            for parent in commit['parents']:
                if parent not in visited:
                    to_visit.append(parent)
        except Exception as e:
            continue

    # Sort by timestamp (newest first)
    def get_timestamp(item):
        sha, commit = item
        author = commit.get('author', '')
        # Parse timestamp from author line
        parts = author.rsplit(' ', 2)
        if len(parts) >= 2:
            try:
                return int(parts[-2])
            except:
                pass
        return 0

    commits.sort(key=get_timestamp, reverse=True)

    # Apply limit
    if limit:
        commits = commits[:limit]

    # Output
    for sha, commit in commits:
        if oneline:
            prefix = '* ' if show_graph else ''
            msg_first_line = commit['message'].split('\n')[0]
            print(f"{prefix}{sha[:7]} {msg_first_line}")
        else:
            if show_graph:
                print("*", end=" ")
            print(f"commit {sha}")

            # Parse author info
            author = commit.get('author', 'Unknown')
            # Format: Name <email> timestamp tz
            parts = author.rsplit(' ', 2)
            if len(parts) >= 3:
                author_info = parts[0]
                timestamp = parts[1]
                tz = parts[2]

                # Format date
                try:
                    ts = int(timestamp)
                    dt = datetime.utcfromtimestamp(ts)
                    date_str = dt.strftime('%a %b %d %H:%M:%S %Y') + f" {tz}"
                except:
                    date_str = f"{timestamp} {tz}"

                print(f"Author: {author_info}")
                print(f"Date:   {date_str}")
            else:
                print(f"Author: {author}")

            print()
            for line in commit['message'].split('\n'):
                print(f"    {line}")
            print()

            # Show stat if requested
            if show_stat and commit['parents']:
                parent_sha = commit['parents'][0]
                try:
                    parent = read_commit(repo_root, parent_sha)
                    show_commit_stat(repo_root, parent['tree'], commit['tree'])
                except:
                    pass

    return 0


def show_commit_stat(repo_root, old_tree_sha: str, new_tree_sha: str):
    """Show file changes between two trees."""
    from objects import read_tree

    def get_tree_files(tree_sha: str, prefix: str = '') -> dict:
        files = {}
        try:
            tree_entries = read_tree(repo_root, tree_sha)
            for entry in tree_entries:
                path = f"{prefix}{entry.name}" if prefix else entry.name
                if entry.mode == '40000':
                    files.update(get_tree_files(entry.sha, path + '/'))
                else:
                    files[path] = entry.sha
        except:
            pass
        return files

    old_files = get_tree_files(old_tree_sha)
    new_files = get_tree_files(new_tree_sha)

    all_paths = set(old_files.keys()) | set(new_files.keys())

    for path in sorted(all_paths):
        old_sha = old_files.get(path)
        new_sha = new_files.get(path)

        if old_sha == new_sha:
            continue

        if old_sha is None:
            print(f" {path} | new file")
        elif new_sha is None:
            print(f" {path} | deleted")
        else:
            print(f" {path} | modified")

File: minigit-3/ts/src/commands/log.ts
// log command - Show commit history

import { findRepoRoot, formatDate, parseTimestamp, shortSha } from '../utils';
import { readObject, parseCommit } from '../objects';
import { resolveRef, getHeadCommit, listBranches } from '../refs';

export function log(args: string[]): void {
  const repoRoot = findRepoRoot();
  if (!repoRoot) {
    console.error('fatal: not a git repository');
    process.exit(1);
  }

  let oneline = false;
  let all = false;
  let limit = Infinity;
  let showGraph = false;
  let showStat = false;
  let revision: string | null = null;

  // Parse arguments
  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    if (arg === '--oneline') {
      oneline = true;
    } else if (arg === '--all') {
      all = true;
    } else if (arg === '--graph') {
      showGraph = true;
    } else if (arg === '--stat') {
      showStat = true;
    } else if (arg === '-n' && i + 1 < args.length) {
      limit = parseInt(args[++i], 10);
    } else if (arg.startsWith('-n')) {
      limit = parseInt(arg.slice(2), 10);
    } else if (!arg.startsWith('-')) {
      revision = arg;
    }
  }

  // Get starting commits
  let startCommits: string[] = [];

  if (all) {
    const branches = listBranches(repoRoot);
    for (const branch of branches) {
      const sha = resolveRef(branch, repoRoot);
      if (sha) startCommits.push(sha);
    }
  } else if (revision) {
    const sha = resolveRef(revision, repoRoot);
    if (!sha) {
      console.error(`fatal: bad revision '${revision}'`);
      process.exit(1);
    }
    startCommits = [sha];
  } else {
    const head = getHeadCommit(repoRoot);
    if (!head) {
      console.error('fatal: your current branch does not have any commits yet');
      process.exit(1);
    }
    startCommits = [head];
  }

  // Walk commits
  const seen = new Set<string>();
  const queue = [...startCommits];
  let count = 0;

  while (queue.length > 0 && count < limit) {
    // Sort by timestamp (most recent first)
    queue.sort((a, b) => {
      try {
        const objA = readObject(a, repoRoot);
        const objB = readObject(b, repoRoot);
        const commitA = parseCommit(objA.content);
        const commitB = parseCommit(objB.content);
        const tsA = parseTimestamp(commitA.committer.split(' ').slice(-2).join(' ')).timestamp;
        const tsB = parseTimestamp(commitB.committer.split(' ').slice(-2).join(' ')).timestamp;
        return tsB - tsA;
      } catch {
        return 0;
      }
    });

    const sha = queue.shift()!;
    if (seen.has(sha)) continue;
    seen.add(sha);

    const obj = readObject(sha, repoRoot);
    if (obj.type !== 'commit') continue;

    const commit = parseCommit(obj.content);

    if (oneline) {
      printOnelineCommit(sha, commit, showGraph);
    } else {
      printFullCommit(sha, commit, showStat, repoRoot);
    }

    count++;

    // Add parents to queue
    for (const parent of commit.parents) {
      if (!seen.has(parent)) {
        queue.push(parent);
      }
    }
  }
}

function printOnelineCommit(sha: string, commit: ReturnType<typeof parseCommit>, showGraph: boolean): void {
  const short = shortSha(sha);
  const message = commit.message.split('\n')[0];
  if (showGraph) {
    console.log(`* ${short} ${message}`);
  } else {
    console.log(`${short} ${message}`);
  }
}

function printFullCommit(sha: string, commit: ReturnType<typeof parseCommit>, showStat: boolean, repoRoot: string): void {
  console.log(`commit ${sha}`);

  if (commit.parents.length > 1) {
    console.log(`Merge: ${commit.parents.map(p => shortSha(p)).join(' ')}`);
  }

  // Parse author info
  const authorMatch = commit.author.match(/^(.+?) <(.+?)> (\d+) ([+-]\d{4})$/);
  if (authorMatch) {
    const [, name, email, timestamp, tz] = authorMatch;
    const dateStr = formatDate(parseInt(timestamp, 10), tz);
    console.log(`Author: ${name} <${email}>`);
    console.log(`Date:   ${dateStr}`);
  } else {
    console.log(`Author: ${commit.author}`);
  }

  console.log('');
  for (const line of commit.message.split('\n')) {
    console.log(`    ${line}`);
  }

  if (showStat && commit.parents.length > 0) {
    printCommitStat(sha, commit.parents[0], repoRoot);
  }

  console.log('');
}

function printCommitStat(commitSha: string, parentSha: string, repoRoot: string): void {
  try {
    const commitObj = readObject(commitSha, repoRoot);
    const parentObj = readObject(parentSha, repoRoot);
    const commit = parseCommit(commitObj.content);
    const parent = parseCommit(parentObj.content);

    const commitFiles = collectFiles(commit.tree, repoRoot);
    const parentFiles = collectFiles(parent.tree, repoRoot);

    const changes: { path: string; added: number; deleted: number }[] = [];

    // Find changes
    for (const [path, sha] of commitFiles) {
      const parentSha = parentFiles.get(path);
      if (!parentSha) {
        // New file
        const size = readObject(sha, repoRoot).size;
        changes.push({ path, added: size, deleted: 0 });
      } else if (parentSha !== sha) {
        // Modified file
        changes.push({ path, added: 1, deleted: 1 });
      }
    }

    for (const [path, sha] of parentFiles) {
      if (!commitFiles.has(path)) {
        // Deleted file
        const size = readObject(sha, repoRoot).size;
        changes.push({ path, added: 0, deleted: size });
      }
    }

    if (changes.length > 0) {
      console.log('');
      for (const change of changes) {
        console.log(` ${change.path}`);
      }
      console.log(` ${changes.length} file${changes.length > 1 ? 's' : ''} changed`);
    }
  } catch {
    // Ignore errors in stat generation
  }
}

function collectFiles(treeSha: string, repoRoot: string, prefix: string = ''): Map<string, string> {
  const { parseTree } = require('../objects');
  const files = new Map<string, string>();
  const obj = readObject(treeSha, repoRoot);
  const entries = parseTree(obj.content);

  for (const entry of entries) {
    const path = prefix ? `${prefix}/${entry.name}` : entry.name;
    if (entry.type === 'tree') {
      for (const [p, s] of collectFiles(entry.sha, repoRoot, path)) {
        files.set(p, s);
      }
    } else {
      files.set(path, entry.sha);
    }
  }

  return files;
}
